mds: add ceph.dir.bal.mask vxattr for MDS Balancer

yongseokoh opened the pull request:
tracker: https://tracker.ceph.com/issues/61777

**Introduction**
This PR introduces the ceph.dir.bal.mask vxattr, which is an option to rebalance a subtree within specific active MDSs. Similar to the CPU mask, this feature enables load balancing of specific directories across multiple MDS ranks. It is especially useful for fine-tuning and improving performance in various scenarios. Previously, the bal_rank_mask in https://github.com/ceph/ceph/pull/43284 supports isolating unpinned subtrees under the root directory ('/') to a specific MDS rank. However, with this new option vxattr, it becomes possible to isolate specific subdirectories to designated MDS ranks. By introducing the ceph.dir.bal.mask vxattr, this PR empowers Ceph administrators with enhanced control and flexibility for optimizing performance and fine-tuning their deployments.

**Use Cases**
The first is when it is difficult to pin a subdir to one MDS rank. The /home/images directory exists. There are /0 to /99 directories under it, and 10 million image files are stored in each directory. In this case, it is difficult to pin the entire images directory to one MDS rank. Also, pinning the huge 100 directories manually or using ephemeral pinning is not an easy task. Therefore, efficient resource management is possible by using ceph.dir.bal.mask.

Second, when there are several large directories such as /home/images, performance can be optimized by distributing them to different MDS rank groups using ceph.dir.bal.mask. Since the existing mdsmap’s bal_rank_mask isolated the entire ‘/’ directory to specific ranks, it can affect performance due to each other's migration overhead. For example, mdsmap’s bal_rank_mask is set to 0xf and /home/images and /home/backups large directories exist. If the load on /home/images instantaneously increases, metadata distribution occurs across ranks 0 to 3. Thus, users of /home/backups may be affected by noisy neighbors unnecessarily. If the two directories are set to MDS rank 0-1 (ceph.dir.bal.mask 0x3) and 2-3 (ceph.dir.bal.mask 0xC) respectively, the effect on each other can be minimized. Like this, it can be used efficiently for various directories.

**How to use**
```
# / root is distributed within MDS rank 0 to 3
# it overrides mdsmap bal_rank_mask value
setfattr -n ceph.dir.bal.mask -v 0x0f /

# /home/images subdir is distributed within MDS rank 1 and 2
setfattr -n ceph.dir.bal.mask -v 0x06 /home/images

# /home/images subdir is distributed within MDS rank 3 and 4
setfattr -n ceph.dir.bal.mask -v 0x18 /home/backups

# remove values and they will obey parent the value or mdsmap bal_rank_mask
setfattr -n ceph.dir.bal.mask -v -1 /home/images
setfattr -n ceph.dir.bal.mask -v -1 /home/backups

# /home/images is placed only on MDS rank1 (similar to ceph.dir.dir)
setfattr -n ceph.dir.bal.mask -v 0x02 /home/images
```


<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

batrick commented on src/mon/FSCommands.cc:
```diff
@@ -432,6 +432,12 @@ class SetHandler : public FileSystemCommandHandler
 	}
+        std::bitset<MAX_MDS> rank_mask = std::bitset<MAX_MDS>(bin_string);
+        if (!rank_mask.test(0)) {
+          ss << "the first bit must be set for rank0";
```
explain (to the user)?

yongseokoh replied:
There have been some slight changes in the implementation. My current implementation involves creating a subtree root for the target directory and exporting that node to a different rank. Then, rebalancing is performed within the masked ranks. For example, if you run "setfattr -n ceph.dir.bal.mask -v 1,2 /home/backup," it converts /home/backup into a subtree and moves it to rank 1. Subsequently, the balancer queries the bal_rank_mask vxattr values of subtrees with authority on that rank and redistributes them within the range of 1 to 2. The '/' root inode is indeed fixed at rank 0, making it impossible to move it to a different rank and apply the logic.  Therefore, I’ve added the constraint to ensure that rank 0 is always included for the root ‘/‘.

The existing bal_rank_mask implementation in https://github.com/ceph/ceph/pull/43284 sends unpinned subtrees (excluding the root '/') to masked ranks for redistribution. The ceph.dir.bal.mask functionality has been implemented, and during the process of integrating the bal_rank_mask feature, constraints were added out of necessity.

batrick reviewed (changes requested):
Few things:

* Please add the PR discussion to your commit message.
* Because the bitset is 256 bits, it's not generally easy to compute the xattr value for the caller. I think it would be helpful to have the MDS compute the value by allowing something like `setfattr -n ceph.dir.bal.mak -v 0,1,3,15 dir/` such that the MDS will do the bitwise or of those bits.
* I'd like to see some tests in `qa/tasks/cephfs/test_exports.py`. You can use `vstart_runner.py` to test.
* There should be some docs added to explain this and the MDSMap bal rank mask. Which should users prefer and when? Is it valid to set the rank mask on the root directory? Are values inherited or override-able?

yongseokoh committed:
test: selftest for ceph.dir.bal.mask vxattr

Signed-off-by: Yongseok Oh <yongseok.oh@linecorp.com>

yongseokoh committed:
doc: document for ceph.dir.bal.mask vxattr

Signed-off-by: Yongseok Oh <yongseok.oh@linecorp.com>

yongseokoh commented on src/include/cephfs/types.h:
```diff
@@ -770,6 +777,7 @@ void inode_t<Allocator>::dump(ceph::Formatter *f) const
   f->dump_int("export_pin", export_pin);
   f->dump_int("export_ephemeral_random_pin", export_ephemeral_random_pin);
   f->dump_bool("export_ephemeral_distributed_pin", export_ephemeral_distributed_pin);
+  f->dump_stream("bal_rank_mask") << bal_rank_mask;
```
replaced with dump_string.

yongseokoh commented on src/mon/FSCommands.cc:
```diff
@@ -432,6 +432,12 @@ class SetHandler : public FileSystemCommandHandler
 	if (r != 0) {
 	  return r;
 	}
+        std::bitset<MAX_MDS> rank_mask = std::bitset<MAX_MDS>(bin_string);
```
I will split the PR once the discussion about restrictionon rank0 is resolved.
https://github.com/ceph/ceph/pull/52373#discussion_r1327957464

yongseokoh commented on src/mds/CInode.cc:
```diff
@@ -2148,13 +2151,16 @@ void CInode::decode_lock_ipolicy(bufferlist::const_iterator& p)
     if (struct_v >= 2) {
       decode(_inode->export_ephemeral_distributed_pin, p);
       decode(_inode->export_ephemeral_random_pin, p);
+      decode(_inode->bal_rank_mask, p);
```
I have added struct_v >= 3 statement for bal_rank_mask.

yongseokoh commented:
@batrick @vshankar  Please review my changes. 
> Few things:
> 
> * Please add the PR discussion to your commit message.

Done.

> * Because the bitset is 256 bits, it's not generally easy to compute the xattr value for the caller. I think it would be helpful to have the MDS compute the value by allowing something like `setfattr -n ceph.dir.bal.mak -v 0,1,3,15 dir/` such that the MDS will do the bitwise or of those bits.

I agree. It is not easy to calculate numerous bits and configure the bitfield.

> * I'd like to see some tests in `qa/tasks/cephfs/test_exports.py`. You can use `vstart_runner.py` to test.

Test cases for ceph.dir.bal.mask were implemented in test_exports.py.

> * There should be some docs added to explain this and the MDSMap bal rank mask. Which should users prefer and when? Is it valid to set the rank mask on the root directory? Are values inherited or override-able?

How to use ceph.dir.bal.mask is explained in the document.

yongseokoh commented:
jenkins retest this please

yongseokoh commented:
jenkins test make check

yongseokoh commented:
jenkins test make check arm64

yongseokoh commented:
@vshankar ping?

vshankar commented:
> @vshankar ping?

Was on vacation. Will recheck tomorrow or early next week.

vshankar commented on src/include/cephfs/types.h:
```diff
@@ -520,6 +521,7 @@ struct inode_t {
 
   double export_ephemeral_random_pin = 0;
   bool export_ephemeral_distributed_pin = false;
+  std::string bal_rank_mask;
```
So, probably the main concern that I have with this change is the increase in the size of the inode which would happen irrespective of the xattr being set.

What are your thoughts on persisting this in the xattr map @yongseokoh ?

yongseokoh replied:
@vshankar There are downsides that may affect inode size. On the other hand, CephFS users have the advantage of being able to flexibly map different subtrees to multiple mds ranks depending on their use case. I think the benefits are great. Also, CephFS users are familiar with using xattr, so I think this method is convenient.

vshankar replied:
That's fine - the xattr interface can remain as it is - just that we could benefit from not increasing the inode size if we can store the mask in xattr_map rather than in the inode.

yongseokoh replied:
Do you mean xattr_map used in https://github.com/ceph/ceph/pull/36603? I will try to see if it can be employed.

yongseokoh replied:
@vshankar 

I have a rough understanding of xattrs through ceph.mirror.dirty_snap_id xattr.

   ``` 
"xattrs": [
         {
             "key": "ceph.mirror.dirty_snap_id",
             "val": "7"
         }
     ]
```

Using xattrs seems like a good idea. However, it would be better for this proposed PR to be managed as the ceph.dir.* family. It seems to be easier to manage in the long term if it is managed in a similar way to the existing ceph.dir.pin and ceph.dir.pin.random. Then, please give your opinion.

vshankar replied:
> Using xattrs seems like a good idea. However, it would be better for this proposed PR to be managed as the ceph.dir.* family. It seems to be easier to manage in the long term if it is managed in a similar way to the existing ceph.dir.pin and ceph.dir.pin.random. Then, please give your opinion.

You do have a point here. I was recommended to use xattrs mirror related xattrs for the sole reason of increasing the inode size. The added member in this change is

> std::string bal_rank_mask;

which when empty and encoded would add (I think) at least 4 bytes to the on-disk size (since the string length needs to be encoded before the actual string). @batrick What do you suggest?

batrick replied:
It would be trivial to add `bal_rank_mask` to the xattr map and add support in the client. As Venky mentioned, see this code:

https://github.com/ceph/ceph/blob/95455490bcdec81906f258e57dbb0fd3e1570e16/src/client/Client.cc#L13808-L13810

https://github.com/ceph/ceph/blob/95455490bcdec81906f258e57dbb0fd3e1570e16/src/client/Client.cc#L14324-L14337

It should be a trivial change. Please make this change @yongseokoh .

yongseokoh replied:
@vshankar @batrick I will change it to use the xattrs map based on your feedback. (It was late because I was on vacation until yesterday. )

vshankar replied:
@yongseokoh Sure thing - thx!

yongseokoh replied:
@vshankar @batrick  I changed this PR to use xattrs map. I confirmed that ceph.dir.bal.mask xattr can be changed through setfattr/getfattr even without changing Client.cc. 
 https://github.com/ceph/ceph/pull/52373/commits/afa6c736b95a1286f9b6b165c5a69068cf8205bf#diff-277dc6e796ccecb6aa14c9357f7a86898d6ddcf4113c110a4e00e0a10f6fefa7R6427

 Do I have to change client.cc as mentioned by Patrick ?

 https://github.com/ceph/ceph/pull/52373#discussion_r1473082207

vshankar replied:
> Do I have to change client.cc as mentioned by Patrick ?

I don't think that's required since the mirror_info xattr is a bit differently handled as it's a compound xattr.

Make sure relevant tests are added for validating get/set of the xattr.

yongseokoh replied:
Yep, tests are included. https://github.com/ceph/ceph/pull/52373/files#diff-b6dd60ae517c65a7733ae0a977eab56298d8f227cbdb4cd0589b8685cbd9669a

vshankar reviewed (commented):
Otherwise the change look reasonable.

vshankar commented:
On 24-01-11 05:01:39, Yongseok Oh wrote:
> @yongseokoh commented on this pull request.
> 
> 
> 
> > @@ -520,6 +521,7 @@ struct inode_t {
>  
>    double export_ephemeral_random_pin = 0;
>    bool export_ephemeral_distributed_pin = false;
> +  std::string bal_rank_mask;
> 
> Do you mean xattr_map used in https://github.com/ceph/ceph/pull/36603? I will try to see if it can be employed.

Yes.

> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/ceph/ceph/pull/52373#discussion_r1448830269
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>

vshankar commented:
jenkins test make check

vshankar commented:
https://pulpito.ceph.com/?branch=wip-vshankar-testing-20240307.022905

Will review the run next week.

vshankar commented on qa/tasks/cephfs/test_exports.py:
```diff
@@ -627,3 +627,372 @@ def test_ephemeral_pin_shrink_mds(self):
+
+        self.mount_a.setfattr("1", "ceph.dir.bal.mask", "2")
+        value = self.mount_a.getfattr("1", "ceph.dir.bal.mask")
+        self.assertEqual(value, "2")
```
The assert check fails - https://pulpito.ceph.com/vshankar-2024-03-07_10:41:12-fs-wip-vshankar-testing-20240307.022905-testing-default-smithi/7584618/

```
2024-03-07T22:51:06.010 INFO:tasks.cephfs_test_runner:======================================================================
2024-03-07T22:51:06.010 INFO:tasks.cephfs_test_runner:FAIL: test_mds_failover (tasks.cephfs.test_exports.TestBalMask)
2024-03-07T22:51:06.010 INFO:tasks.cephfs_test_runner:That MDS failover does not affect the ceph.dir.bal.mask.
2024-03-07T22:51:06.010 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2024-03-07T22:51:06.011 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2024-03-07T22:51:06.011 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_b0ea271bb96f7f4cec2f2a40ff88a5def1e0530f/qa/tasks/cephfs/test_exports.py", line 934, in test_mds_failover
2024-03-07T22:51:06.011 INFO:tasks.cephfs_test_runner:    self.assertEqual(value, "2")
2024-03-07T22:51:06.011 INFO:tasks.cephfs_test_runner:AssertionError: None != '2'
```

yongseokoh replied:
@vshankar  I will check and get back to you.

yongseokoh replied:
In the vstart-based local test, all 20 test cases passed. I will look at the log you provided and see if I can reproduce it.

```
2024-03-13 11:38:19,031.031 INFO:__main__:test_unset_rank_mask_under_nested_parent (tasks.cephfs.test_exports.TestBalMask) ... ok
2024-03-13 11:38:19,031.031 INFO:__main__:test_unset_rank_mask_under_nested_root (tasks.cephfs.test_exports.TestBalMask) ... ok
2024-03-13 11:38:19,031.031 INFO:__main__:
2024-03-13 11:38:19,031.031 INFO:__main__:----------------------------------------------------------------------
2024-03-13 11:38:19,031.031 INFO:__main__:Ran 20 tests in 1378.896s
2024-03-13 11:38:19,031.031 INFO:__main__:
2024-03-13 11:38:19,032.032 INFO:__main__:
2024-03-13 11:38:19,032.032 INFO:__main__:> ip netns list
2024-03-13 11:38:19,039.039 INFO:__main__:> sudo ip link delete ceph-brx
Cannot find device "ceph-brx"
2024-03-13 11:38:19,066.066 INFO:__main__:OK
2024-03-13 11:38:19,067.067 INFO:__main__:
```

yongseokoh replied:
@vshankar 
I think it's a timing issue. This is the case where getfattr fails after setfattr regardless of MDS fail. Can MDS logs be shared?

yongseokoh replied:
The major difference is the use of the kernel client. I only tested ceph-fuse.

yongseokoh replied:
The setxattr client request was processed and a getxattr request was requested between early_reply and reply_client_request. I think the client inode xattrs was not updated at that time. Is it possible to guarantee that the client inode's xattrs will be updated after early_reply?

```
2024-03-07T22:50:52.520+0000 7f8607400640  4 mds.0.server handle_client_request client_request(client.5210:27 setxattr #0x10000000000 ceph.dir.bal.mask 2024-03-07T22:50:52.519761+0000 caller_uid=1000, caller_gid=1301{6,36,1000,1301,}) v6
2024-03-07T22:50:52.521+0000 7f8607400640 10 mds.0.server setxattr 'ceph.dir.bal.mask' len 1 on [inode 0x10000000000 [...2,head] /1/ auth v4 ap=2 DIRTYPARENT f(v0 m2024-03-07T22:50:51.975767+0000 1=0+1) n(v0 rc2024-03-07T22:50:51.975767+0000 2=0+2) (isnap sync r=1) (inest lock) (ifile excl) (ixattr xlock x=1 by 0x55c05e997a80) (iversion lock w=1 last_client=5210) caps={5210=pAsLsFsx/pAsLsXsFsxcral@2},l=5210 | request=1 lock=3 dirfrag=1 caps=1 dirtyparent=1 dirty=1 authpin=1 0x55c05e90a100]
2024-03-07T22:50:52.521+0000 7f8607400640 10 mds.0.server early_reply 0 ((0) Success) client_request(client.5210:27 setxattr #0x10000000000 ceph.dir.bal.mask 2024-03-07T22:50:52.519761+0000 caller_uid=1000, caller_gid=1301{6,36,1000,1301,}) v6


2024-03-07T22:50:52.580+0000 7f8607400640  4 mds.0.server handle_client_request client_request(client.5210:28 ??? #0x10000000000 ceph.dir.bal.mask 2024-03-07T22:50:52.578761+0000 caller_uid=1000, caller_gid=1301{6,36,1000,1301,}) v6
2024-03-07T22:50:52.580+0000 7f8607400640  7 mds.0.server reply_client_request -61 ((61) No data available) client_request(client.5210:28 ??? #0x10000000000 ceph.dir.bal.mask 2024-03-07T22:50:52.578761+0000 caller_uid=1000, caller_gid=1301{6,36,1000,1301,}) v6

inode->xattrs may be not updated.
https://github.com/ceph/ceph/blob/main/src/client/Client.cc#L13757

2024-03-07T22:50:55.379+0000 7f85ff200640  7 mds.0.server reply_client_request 0 ((0) Success) client_request(client.5210:27 setxattr #0x10000000000 ceph.dir.bal.mask 2024-03-07T22:50:52.519761+0000 caller_uid=1000, caller_gid=1301{6,36,1000,1301,}) v6
```

vshankar replied:
> The setxattr client request was processed and a getxattr request was requested between early_reply and reply_client_request. I think the client inode xattrs was not updated at that time. Is it possible to guarantee that the client inode's xattrs will be updated after early_reply? Otherwise, should I just add some logic in the test case to wait for a few seconds?

If the client has the relevant caps, it should have the latest xattr version. I think you might have run into a race where the client thinks it has the caps and can trust the cached xattr, however, the xattr is stale.

I have seen something similar before and fixed it in a code path. See commit 526e4ee5b4b7c27306b70de62daa5cf705763130

> 
> ```
> 2024-03-07T22:50:52.520+0000 7f8607400640  4 mds.0.server handle_client_request client_request(client.5210:27 setxattr #0x10000000000 ceph.dir.bal.mask 2024-03-07T22:50:52.519761+0000 caller_uid=1000, caller_gid=1301{6,36,1000,1301,}) v6
> 2024-03-07T22:50:52.521+0000 7f8607400640 10 mds.0.server setxattr 'ceph.dir.bal.mask' len 1 on [inode 0x10000000000 [...2,head] /1/ auth v4 ap=2 DIRTYPARENT f(v0 m2024-03-07T22:50:51.975767+0000 1=0+1) n(v0 rc2024-03-07T22:50:51.975767+0000 2=0+2) (isnap sync r=1) (inest lock) (ifile excl) (ixattr xlock x=1 by 0x55c05e997a80) (iversion lock w=1 last_client=5210) caps={5210=pAsLsFsx/pAsLsXsFsxcral@2},l=5210 | request=1 lock=3 dirfrag=1 caps=1 dirtyparent=1 dirty=1 authpin=1 0x55c05e90a100]
> 2024-03-07T22:50:52.521+0000 7f8607400640 10 mds.0.server early_reply 0 ((0) Success) client_request(client.5210:27 setxattr #0x10000000000 ceph.dir.bal.mask 2024-03-07T22:50:52.519761+0000 caller_uid=1000, caller_gid=1301{6,36,1000,1301,}) v6
> 
> 
> 2024-03-07T22:50:52.580+0000 7f8607400640  4 mds.0.server handle_client_request client_request(client.5210:28 ??? #0x10000000000 ceph.dir.bal.mask 2024-03-07T22:50:52.578761+0000 caller_uid=1000, caller_gid=1301{6,36,1000,1301,}) v6
> 2024-03-07T22:50:52.580+0000 7f8607400640  7 mds.0.server reply_client_request -61 ((61) No data available) client_request(client.5210:28 ??? #0x10000000000 ceph.dir.bal.mask 2024-03-07T22:50:52.578761+0000 caller_uid=1000, caller_gid=1301{6,36,1000,1301,}) v6
> 
> inode->xattrs may be not updated.
> https://github.com/ceph/ceph/blob/main/src/client/Client.cc#L13757
> 
> 2024-03-07T22:50:55.379+0000 7f85ff200640  7 mds.0.server reply_client_request 0 ((0) Success) client_request(client.5210:27 setxattr #0x10000000000 ceph.dir.bal.mask 2024-03-07T22:50:52.519761+0000 caller_uid=1000, caller_gid=1301{6,36,1000,1301,}) v6
> ```

vshankar reviewed (changes requested):
@yongseokoh PTAL at the failed tests.

As far as this run in concerned, we ran into a bug with the testing kernel causing some tests failures unrelated to this change. See - https://tracker.ceph.com/issues/64679#note-4

Once those two issues are fixed, the failed tests would need a rerun. Other test look good 👍 

This is getting really close to getting merged.

vshankar commented:
> https://pulpito.ceph.com/?branch=wip-vshankar-testing-20240307.022905
> 
> Will review the run next week.

Jusy FYI - apart from one failures reported here, there is an unrelated test case failure due to a bug in the testing kernel.

batrick commented on qa/tasks/cephfs/test_exports.py:
```diff
@@ -627,3 +627,372 @@ def test_ephemeral_pin_shrink_mds(self):
+        self.assertEqual(value, None)
+
+        self.mount_a.setfattr(".", "ceph.dir.bal.mask", "0")
+        self._wait_subtrees([('', 0)], status=self.status, path='', exclude_path='~mds')
```
I'm not sure what this test is doing that `test_unset_rank_mask` isn't already testing?

yongseokoh replied:
This unset test may be redundant, but '.' The directory is located at rank 0, and the '1' directory is exported at rank 2. The purpose is to test whether the parent directory's export affects the '1' directory.

yongseokoh replied:
It seems the same line. https://github.com/ceph/ceph/pull/52373#discussion_r1532584706

batrick commented on qa/tasks/cephfs/test_exports.py:
```diff
@@ -627,3 +627,372 @@ def test_ephemeral_pin_shrink_mds(self):
+
+        self.fs.set_max_mds(2)
+        self.status = self.fs.wait_for_daemons()
+        self._wait_subtrees([('/1', 0)], status=self.status)
```
why can it not be on rank 1? Is there a documented rule for this scenario?

yongseokoh replied:
I understand that when max_mds shrink occurs, the subtrees in mds move to rank 0. 
https://github.com/ceph/ceph/blob/main/src/mds/MDCache.cc#L7909

Well, if there is a point that I don't understand, please leave a comment.

batrick replied:
I see. Please make sure the documentation addresses this.

yongseokoh replied:
Added.

https://github.com/ceph/ceph/compare/9b04cd83fb615464888de665b9933d37724d4a28..ac1ab14f09a789361f9a5c72e448ae85f5b55fa8#diff-cc9cf0d49df0c5644ecb9c98d1f7ac1d307b05837bfdabc6761c65691c6f8dc1R438

yongseokoh commented on doc/cephfs/multimds.rst:
```diff
@@ -311,3 +311,114 @@ the ``bal_rank_mask`` should be set to ``0x0``. For example:
+
+If the ``ceph.dir.bal.mask`` for the root ``/`` directory is set to ``0,1,2,3,4,5``,
+the ``bal_rank_mask`` will be overridden. In this way, the balancer dynamically
+distributes the unpinned subtrees of the root ``/`` directory from ``0`` to ``5`` mds ranks.
```
```bal_rank_mask``` does not need to be deprecated. If detailed tuning is required according to the user's preference, ```ceph.dir.bal.mask``` xattr can be used. Otherwise, ```bal_rank_mask``` can be used for simplicity.

batrick replied:
Perhaps add a note about that, literally:

```
.. note:: Users may prefer to use the file system setting `bal_rank_mask` ...
```

yongseokoh replied:
Added. https://github.com/ceph/ceph/compare/9b04cd83fb615464888de665b9933d37724d4a28..ac1ab14f09a789361f9a5c72e448ae85f5b55fa8#diff-cc9cf0d49df0c5644ecb9c98d1f7ac1d307b05837bfdabc6761c65691c6f8dc1R407

yongseokoh commented on doc/cephfs/multimds.rst:
```diff
@@ -311,3 +311,114 @@ the ``bal_rank_mask`` should be set to ``0x0``. For example:
+for a single MDS rank to handle efficiently. For example, the ``/home`` directory
+contains many directories for multiple users. In most cases, static pinning is
+simply used to distribute subdirectories of /home across multiple ranks. However,
+the ``/home/images`` directory presents a unique situation with subdirectories 0 to 99,
```
I think ```/usr``` is better. However, I don't think it will mount cephfs on ```/usr```. What about examples like ```/data``` or ```/cephfs``` directories?

batrick replied:
Well, how do you plan to use it? I think `/usr` could make sense in the context of a backing store of a large software respository file system like cvmfs. It just requires exposition to explain.

yongseokoh replied:
We need to find a good use case. We had an issue with saving large files at https://laion.ai/blog/laion-400-open-dataset. In fact, cephfs users mount and store file systems in any directory. It is difficult to determine a general case because the user stores it in any directory they want. For example: ```/home/$user/laion-5b```, ```/laion-5b``` , or ```/mnt/laion-5b```. cvmfs is also a good example. I'll think about it.

yongseokoh replied:
There are no good examples. cvmfs also uses the /cvmfs directory. some helpful explanation is added to the note.

https://github.com/ceph/ceph/compare/9b04cd83fb615464888de665b9933d37724d4a28..ac1ab14f09a789361f9a5c72e448ae85f5b55fa8#diff-cc9cf0d49df0c5644ecb9c98d1f7ac1d307b05837bfdabc6761c65691c6f8dc1R344

yongseokoh commented on qa/tasks/cephfs/test_exports.py:
```diff
@@ -627,3 +627,372 @@ def test_ephemeral_pin_shrink_mds(self):
+                    hit_ranks.add(rank)
+            time.sleep(5)
+
+        self.assertEqual([1,2], sorted(list(hit_ranks)))
```
Applying ```ceph.dir.bal.mask``` to one rank has the same effect as ```ceph.dir.pin```. In the example above, the ```dir/``` directory is moved to rank ```1``` and the balanacer only operates at rank ```1```, so no additional subtree movement occurs. If there is confusion with ```ceph.dir.pin```, how about limiting the use of ```ceph.dir.bal.mask``` to 2 or more ranks?

batrick replied:
No, let's allow it if for no other reason than it makes programs modifying the mask easier.

I think just adding a note in the docs about this would be helpful.

But my question still is: should we get rid of the special code handling static export pins and have `ceph.dir.pin` translate to `ceph.dir.bal.mask`?

batrick replied:
Follow-up question: does the new `balance_automate` setting disable `ceph.dir.bal.mask`? Should it?

yongseokoh replied:
Your idea is that we should keep only the interface of ```ceph.dir.pin``` and integrate the related functions into `ceph.dir.bal.mask`? It's possible, but there may be other things to consider. When only the value ```1``` is stored, if getattr is called through two interfaces, both normally return ```1```. The values ```1,2``` are then stored through ceph.dir.bal.mask. After that, when reading `ceph.dir.pin`, an invalid value error is returned, and when reading` ceph.dir.bal.mask`, ```1,2``` is returned normally. Users may be confused. (If I misunderstand your idea, please comment again.)

Since `ceph.dir.pin` is a function that is very familiar to Ceph users, it would be better to keep it as is. I think it would be good to mention the similarities and differences with` ceph.dir.pin `in the documentation.

In the current implementation, subtree migration by updating `ceph.dir.bal.mask` does not occur when Balance_automate is false. Users may want to stop the migration in some cases, so that feature should be provided.

batrick replied:
> Your idea is that we should keep only the interface of `ceph.dir.pin` and integrate the related functions into `ceph.dir.bal.mask`? It's possible, but there may be other things to consider. When only the value `1` is stored, if getattr is called through two interfaces, both normally return `1`. The values `1,2` are then stored through ceph.dir.bal.mask. After that, when reading `ceph.dir.pin`, an invalid value error is returned, and when reading` ceph.dir.bal.mask`, `1,2` is returned normally. Users may be confused. (If I misunderstand your idea, please comment again.)

I think that particular issue boils down to whether we should allow both `ceph.dir.pin` and `ceph.dir.bal.mask` on the same inode. I think you could keep `ceph.dir.pin` metadata in the `inode_t` but actually implement it in the backend as a `ceph.dir.bal.mask`.

Anyway, it's just a thought that could be refactored later perhaps.

> Since `ceph.dir.pin` is a function that is very familiar to Ceph users, it would be better to keep it as is. I think it would be good to mention the similarities and differences with`ceph.dir.pin`in the documentation.

Yes, I don't think we would retire the interface.
 
> In the current implementation, subtree migration by updating `ceph.dir.bal.mask` does not occur when Balance_automate is false. Users may want to stop the migration in some cases, so that feature should be provided.

Please add a documentation note.

yongseokoh replied:
Added.

https://github.com/ceph/ceph/compare/9b04cd83fb615464888de665b9933d37724d4a28..ac1ab14f09a789361f9a5c72e448ae85f5b55fa8#diff-cc9cf0d49df0c5644ecb9c98d1f7ac1d307b05837bfdabc6761c65691c6f8dc1R424

yongseokoh commented on qa/tasks/cephfs/test_exports.py:
```diff
@@ -627,3 +627,372 @@ def test_ephemeral_pin_shrink_mds(self):
+        """
+
+        self.mount_a.setfattr(".", "ceph.dir.bal.mask", "0")
+        self._wait_subtrees([('', 0)], status=self.status, path='', exclude_path='~mds')
```
Fixed.

yongseokoh commented on qa/tasks/cephfs/cephfs_test_case.py:
```diff
@@ -328,7 +328,7 @@ def delete_mds_coredump(self, daemon_id):
             log.info("No core_pattern directory set, nothing to clear (internal.coredump not enabled?)")
 
-    def _get_subtrees(self, status=None, rank=None, path=None):
+    def _get_subtrees(self, status=None, rank=None, path=None, exclude_path=None):
```
Fixed.

yongseokoh commented on qa/tasks/cephfs/cephfs_test_case.py:
```diff
@@ -354,14 +356,15 @@ def _get_subtrees(self, status=None, rank=None, path=None):
+                    subtrees = self._get_subtrees(status=status, rank=rank, path=path, exclude_path=exclude_path)
+                    log.info(subtrees)
                     filtered = sorted([(s['dir']['path'], s['auth_first']) for s in subtrees])
-                    log.info("%s =?= %s", filtered, test)
```
Reverted.

yongseokoh commented on qa/tasks/cephfs/test_exports.py:
```diff
@@ -627,3 +627,372 @@ def test_ephemeral_pin_shrink_mds(self):
+        """
+        That MDS failover does not affect the ceph.dir.bal.mask.
+        """
+        self.config_set('mds', 'mds_debug_subtrees', False)
```
It was deleted as it was unnecessary.

yongseokoh commented on doc/cephfs/multimds.rst:
```diff
@@ -311,3 +311,114 @@ the ``bal_rank_mask`` should be set to ``0x0``. For example:
+Dynamically partitionning directory trees with vxattr on a specific ranks
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The ``bal_rank_mask`` option can be overridden by configuring the ``ceph.dir.bal.mask`` vxattr.
```
Fixed

yongseokoh commented on doc/cephfs/multimds.rst:
```diff
@@ -311,3 +311,114 @@ the ``bal_rank_mask`` should be set to ``0x0``. For example:
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The ``bal_rank_mask`` option can be overridden by configuring the ``ceph.dir.bal.mask`` vxattr.
+Notably, the ``ceph.dir.bal.mask`` offers the capability to segregate particular subdirectories
```
Fixed.

yongseokoh commented on doc/cephfs/multimds.rst:
```diff
@@ -311,3 +311,114 @@ the ``bal_rank_mask`` should be set to ``0x0``. For example:
+The ``bal_rank_mask`` option can be overridden by configuring the ``ceph.dir.bal.mask`` vxattr.
+Notably, the ``ceph.dir.bal.mask`` offers the capability to segregate particular subdirectories
+into dedicated groups of MDS ranks, whereas bal_rank_mask only allows isolation
+at the filesystem level, such as the root directory ``/``. It serves as a valuable
```
Fixed.

yongseokoh commented on doc/cephfs/multimds.rst:
```diff
@@ -311,3 +311,114 @@ the ``bal_rank_mask`` should be set to ``0x0``. For example:
+The ``ceph.dir.bal.mask`` can also be a useful option for fine-tunning performance,
+particularly in scenarios involving large directories like ``/home/images``
+and ``/home/backups`` within the file system.
+While the mdsmap’s ``bal_rank_mask`` isolates the entire ``/`` directory to
```
Fixed.

yongseokoh commented on doc/cephfs/multimds.rst:
```diff
@@ -311,3 +311,114 @@ the ``bal_rank_mask`` should be set to ``0x0``. For example:
+
+Similar to ``ceph.dir.pin``, the ``ceph.dir.bal.mask`` is also inherited from its closest parent.
+This manner involves configuring the ``ceph.dir.bal.mask`` for a directory, which consequently
+impacts all its subordinate items. However, the parent ``ceph.dir.bal.mask`` can be overriden
```
Fixed.

yongseokoh commented on doc/cephfs/multimds.rst:
```diff
@@ -311,3 +311,114 @@ the ``bal_rank_mask`` should be set to ``0x0``. For example:
+
+    mkdir -p /home/images/jpg
+    setfattr -n ceph.dir.bal.mask -v 1,2 /home/images
+    # /home/images and /home/images/jpg are now freely distributed within mds rank 1 and 2
```
Fixed.

yongseokoh commented on doc/cephfs/multimds.rst:
```diff
@@ -311,3 +311,114 @@ the ``bal_rank_mask`` should be set to ``0x0``. For example:
+    setfattr -n ceph.dir.bal.mask -v 1,2 /home/images
+    # /home/images and /home/images/jpg are now freely distributed within mds rank 1 and 2
+    setfattr -n ceph.dir.bal.mask -v 3,4 /home/images/jpg
+    # /home/images/jpg are distributed within mds rand 3 and 4
```
Fixed.

yongseokoh commented on doc/cephfs/multimds.rst:
```diff
@@ -311,3 +311,114 @@ the ``bal_rank_mask`` should be set to ``0x0``. For example:
+
+::
+
+    setfattr -n ceph.dir.bal.mask -v -1 /home/images/jpg
```
Fixed.

yongseokoh commented on doc/cephfs/multimds.rst:
```diff
@@ -311,3 +311,114 @@ the ``bal_rank_mask`` should be set to ``0x0``. For example:
+    setfattr -n ceph.dir.bal.mask -v 2,3 /home/images/jpg
+
+The ``/home/images`` directory will be pinned to rank 1, while the ``/home/images/jpg``
+directory will be dynamically split into rank 2 and 3.
```
Fixed.

yongseokoh commented on doc/cephfs/multimds.rst:
```diff
@@ -311,3 +311,114 @@ the ``bal_rank_mask`` should be set to ``0x0``. For example:
+The ``/home/images`` directory will be pinned to rank 1, while the ``/home/images/jpg``
+directory will be dynamically split into rank 2 and 3.
+
+**Restrictions**: Since the inode of the root directory is anchored in MDS rank0,
```
Fixed.

yongseokoh commented on doc/cephfs/multimds.rst:
```diff
@@ -311,3 +311,114 @@ the ``bal_rank_mask`` should be set to ``0x0``. For example:
+directory will be dynamically split into rank 2 and 3.
+
+**Restrictions**: Since the inode of the root directory is anchored in MDS rank0,
+it's reuiqred to include MDS rank 0 when adjusting the ceph.dir.rank.mask for the ``/`` directory.
```
Fixed.

yongseokoh commented on qa/tasks/cephfs/test_exports.py:
```diff
@@ -627,3 +627,372 @@ def test_ephemeral_pin_shrink_mds(self):
+
+        hit_ranks = set()
+        for rank in [0, 1, 2]:
+            subtrees = self._get_subtrees(status=self.status, rank=rank)
```
Fixed.

yongseokoh commented on qa/tasks/cephfs/test_exports.py:
```diff
@@ -627,3 +627,372 @@ def test_ephemeral_pin_shrink_mds(self):
+            subtrees = self._get_subtrees(status=self.status, rank=rank)
+            for s in subtrees:
+                path = s['dir']['path']
+                if path.startswith('/1') and s['auth_first'] == rank:
```
Fixed.

yongseokoh commented on qa/tasks/cephfs/cephfs_test_case.py:
```diff
@@ -328,7 +328,7 @@ def delete_mds_coredump(self, daemon_id):
-    def _get_subtrees(self, status=None, rank=None, path=None):
+    def _get_subtrees(self, status=None, rank=None, path=None, exclude_path=None):
         if path is None:
             path = "/"
```
Changed.

yongseokoh commented on qa/tasks/cephfs/cephfs_test_case.py:
```diff
@@ -344,6 +344,8 @@ def _get_subtrees(self, status=None, rank=None, path=None):
                             subtrees = self.fs.rank_asok(["get", "subtrees"], status=status, rank=rank)
                         subtrees = filter(lambda s: s['dir']['path'].startswith(path), subtrees)
+                        if exclude_path:
+                            subtrees = filter(lambda s: not s['dir']['path'].startswith(exclude_path), subtrees)
```
Added.

yongseokoh commented on qa/tasks/cephfs/test_exports.py:
```diff
@@ -627,3 +627,365 @@ def test_ephemeral_pin_shrink_mds(self):
+        self.assertEqual(value, None)
+
+        self.mount_a.setfattr(".", "ceph.dir.bal.mask", "0")
+        self._wait_subtrees([('', 0)], status=self.status, path='')
```
Added comments.

https://github.com/ceph/ceph/compare/9b04cd83fb615464888de665b9933d37724d4a28..ac1ab14f09a789361f9a5c72e448ae85f5b55fa8#diff-b6dd60ae517c65a7733ae0a977eab56298d8f227cbdb4cd0589b8685cbd9669aR904

yongseokoh commented on src/mds/MDCache.cc:
```diff
@@ -982,8 +993,15 @@ void MDCache::try_subtree_merge_at(CDir *dir, set<CInode*> *to_eval, bool adjust
+  int r = mds->balancer->get_rank_mask_bitset(dir, rank_mask_bitset, false);
+  bool has_mask = false;
+  if (r == 0 && rank_mask_bitset.count()) {
+    has_mask = false;
```
false to true

yongseokoh commented on src/mds/MDBalancer.cc:
```diff
@@ -186,6 +240,7 @@ void MDBalancer::handle_export_pins(void)
          */
         if (dir->get_num_head_items() > 0) {
 	  mds->mdcache->migrator->export_dir(dir, target);
+        } else {
```
unnecessary

batrick replied:
remove?

yongseokoh commented:
@batrick  ping? Could you please review this PR when you get a chance?

batrick commented on doc/cephfs/multimds.rst:
```diff
@@ -311,3 +313,151 @@ the ``bal_rank_mask`` should be set to ``0x0``. For example:
+contains many directories for multiple users. In most cases, static pinning is
+simply used to distribute subdirectories of ``/usr`` across multiple ranks. However,
+the ``/usr/share`` directory presents a unique situation with many subdirectories,
+each containing large number of files, often exceeding a million files. With
+this example with ``/usr/share/images``, processing them in one MDS is inappropriate
+due to insufficient MDS resources like MDS cache memory. Therefore,
+the ceph.dir.bal.mask can dynamically balance the workload of this large
+directory within a specific subset of ranks.
```
```suggestion
each containing large number of files, often exceeding a million files. For
example, consider``/usr/share/images`` with several million images. Processing this directory on one MDS may be inappropriate
due to insufficient MDS resources like MDS cache memory. Therefore,
the ceph.dir.bal.mask could dynamically balance the workload of this large
directory within a specific subset of ranks.
```

batrick commented on doc/cephfs/multimds.rst:
```diff
@@ -311,3 +313,151 @@ the ``bal_rank_mask`` should be set to ``0x0``. For example:
+    setfattr -n ceph.dir.bal.mask -v 1,2 /usr/share/images
+    setfattr -n ceph.dir.bal.mask -v 3,4 /usr/share/backups
+
+.. note:: The balance_automate setting should be `true` for this option to be turned on.
```
```suggestion
.. note:: The balance_automate setting must be `true` for `ceph.dir.bal.mask` to have any effect.
```

batrick commented on doc/cephfs/multimds.rst:
```diff
@@ -311,3 +313,151 @@ the ``bal_rank_mask`` should be set to ``0x0``. For example:
+its distinct advantage lies in its flexibility.
+Unlike `ceph.dir.pin`, `ceph.dir.bal.mask` can be expanded to multiple mds ranks or reduced to fewer mds ranks.
+Therefore, in scenarios requiring expansion to multiple mds ranks,
+it is recommended to employ `ceph.dir.bal.mask`.
```
On the other hand, if the admin wants the automated balancer to be off, then `ceph.dir.pin` or ephemeral pinning is the way to distribute load.

batrick commented on doc/cephfs/multimds.rst:
```diff
@@ -311,3 +313,151 @@ the ``bal_rank_mask`` should be set to ``0x0``. For example:
+
+    # The subtree is now moved to mds rank 0.
+
+**Restrictions**: Since the inode of the root directory is anchored to MDS rank 0,
```
```suggestion
**Restrictions**: Since the inode of the root directory is always implicitly pinned to MDS rank 0,
```

batrick reviewed (changes requested):
Please also reorg the commits:

- Feature
- Tests
- Docs

batrick commented:
(sorry for the delay!)

batrick commented:
> @batrick I made changes based on your comments. This PR has already been categorized into 3 commits. Should I do something more?
> 
> > Please also reorg the commits:
> > Feature
> > Tests
> > Docs

I think it was reversed before but now it looks right.

batrick commented on src/mds/CInode.cc:
```diff
@@ -2112,6 +2117,9 @@ void CInode::decode_lock_ixattr(bufferlist::const_iterator& p)
   }
   DECODE_FINISH(p);
   reset_inode(std::move(_inode));
+
+  std::string bal_rank_mask = get_bal_rank_mask_from_xattrs(false);
+  maybe_export_pin(prev_bal_rank_mask != bal_rank_mask);
```
Let's accept this as-is, since we asked you to do it this way; but, I plan to restructure this to use the policylock (which is the right lock). I don't like putting this information in the xattrmap as it relies on using the cap-related locks and will introduce numerous similar checks for every ixattr decode. It's also simply incoherent with the current design of locks.

My plan is to create another lightweight map in the CInode which can take any number of these rarely set metadata values so it's efficient to encode/decode. The in-memory version would be a std::vector for efficient lookup (a map is unnecessary for small, and often empty, datasets).

If you want to make that change before merging this please let me know. Otherwise, I'll do it post-merge.

yongseokoh replied:
I think developing a new lightweight map for CInode looks very interesting. However, it would be better to proceed separately from this PR. I prefer to merge this PR first.

batrick commented on src/mds/MDBalancer.h:
```diff
@@ -75,6 +76,10 @@ class MDBalancer {
   void handle_mds_failure(mds_rank_t who);
 
   int dump_loads(Formatter *f, int64_t depth = -1) const;
+  int get_rank_mask_bitset(CDir *dir, std::bitset<MAX_MDS>& rank_mask_bitset, bool inherit=true);
```
Let's typedef `std::bitset<MAX_MDS>`. Put it in `mdstypes.h`.

yongseokoh replied:
Defined using typedef in `mdstypes.h`

batrick reviewed (changes requested):
small nit otherwise LGTM

yongseokoh commented on src/mds/MDBalancer.h:
```diff
@@ -75,6 +76,10 @@ class MDBalancer {
 
   int dump_loads(Formatter *f, int64_t depth = -1) const;
+  int get_rank_mask_bitset(CDir *dir, max_mds_bitset_t& rank_mask_bitset, bool inherit=true);
+  std::string bitmask_to_str(std::bitset<MAX_MDS> &bitmask);
```
I made a mistake. I modified them. Thank you.

batrick commented:
jenkins test make check arm64

batrick commented:
This PR is under test in [https://tracker.ceph.com/issues/67214](https://tracker.ceph.com/issues/67214).

batrick reviewed (changes requested):
Saw this:

https://pulpito.ceph.com/pdonnell-2024-07-29_20:36:30-fs-wip-pdonnell-testing-20240726.202642-debug-distro-default-smithi/7824827/

PTAL. @yongseokoh do you have sepia lab access? Please see https://wiki.sepia.ceph.com/doku.php?id=vpnaccess if not.

yongseokoh commented:
@batrick  I'll check in a few days. I need some time.

> Saw this:
> 
> https://pulpito.ceph.com/pdonnell-2024-07-29_20:36:30-fs-wip-pdonnell-testing-20240726.202642-debug-distro-default-smithi/7824827/
> 
> PTAL. @yongseokoh do you have sepia lab access? Please see https://wiki.sepia.ceph.com/doku.php?id=vpnaccess if not.

yongseokoh commented:
@batrick There was a bug in CInode::get_rank_mask_inode(), so I fixed it. Some of the numbers were changed because a timeout occurred during the process of creating multiple files in some of the test cases. All the tests I developed passed.

```
2024-08-16 20:13:56,258.258 INFO:__main__:test_unset_rank_mask_under_nested_parent (tasks.cephfs.test_exports.TestBalMask.test_unset_rank_mask_under_nested_parent)
2024-08-16 20:13:56,258.258 INFO:__main__:That ceph.dir.bal.mask is unset under nested directory. ... ok
2024-08-16 20:13:56,258.258 INFO:__main__:test_unset_rank_mask_under_nested_root (tasks.cephfs.test_exports.TestBalMask.test_unset_rank_mask_under_nested_root)
2024-08-16 20:13:56,258.258 INFO:__main__:That ceph.dir.bal.mask is unset under root directory. ... ok
2024-08-16 20:13:56,258.258 INFO:__main__:
2024-08-16 20:13:56,258.258 INFO:__main__:----------------------------------------------------------------------
2024-08-16 20:13:56,258.258 INFO:__main__:
2024-08-16 20:13:56,259.259 INFO:__main__:> ip netns list
2024-08-16 20:13:56,262.262 INFO:__main__:> sudo ip link delete ceph-brx
Cannot find device "ceph-brx"
2024-08-16 20:13:56,281.281 INFO:__main__:Ran 20 tests in 1302.834s
2024-08-16 20:13:56,281.281 INFO:__main__:
2024-08-16 20:13:56,281.281 INFO:__main__:OK
2024-08-16 20:13:56,282.282 INFO:__main__:
```

yongseokoh commented:
jenkins test make check

batrick commented:
I'll try again, thanks.

batrick commented on src/mds/CInode.cc:
```diff
@@ -5492,6 +5512,59 @@ void CInode::setxattr_ephemeral_dist(bool val)
+
+CInode *CInode::get_rank_mask_inode(bool inherit)
+{
+  if (!g_conf().get_val<bool>("mds_bal_export_pin"))
```
This will be too expensive to run so frequently. Please cache the config variable in MDCache (as we do elsewhere).

yongseokoh replied:
@batrick I fixed it.

yongseokoh commented:
jenkins test make check

yongseokoh commented:
jenkins test api

yongseokoh commented:
jenkins test make check

yongseokoh commented:
jenkins test make check

yongseokoh commented:
@batrick Could you please review this updated PR when you get a moment?

anthonyeleven commented on doc/cephfs/multimds.rst:
```diff
@@ -314,3 +313,151 @@ the ``bal_rank_mask`` should be set to ``0x0``. For example:
+
+.. note:: The example here uses the /usr/share directory.
+        Depending on the user's preference, it can be applied
+        to various directories such as /usr/share and /mnt/cephfs.
```
double backticks?

anthonyeleven commented on doc/cephfs/multimds.rst:
```diff
@@ -314,3 +313,151 @@ the ``bal_rank_mask`` should be set to ``0x0``. For example:
+impacts all its descendants. However, the parent ``ceph.dir.bal.mask`` can be overridden
+by setting the child directory’s value. For example:
+
+::
```
@bluikko Ditto wrt this block and those below

anthonyeleven commented on doc/cephfs/multimds.rst:
```diff
@@ -314,3 +313,151 @@ the ``bal_rank_mask`` should be set to ``0x0``. For example:
+its distinct advantage lies in its flexibility.
+Unlike `ceph.dir.pin`, `ceph.dir.bal.mask` can be expanded to multiple mds ranks or reduced to fewer mds ranks.
+Therefore, in scenarios requiring expansion to multiple mds ranks,
+it is recommended to employ `ceph.dir.bal.mask`.
```
MDS uppercase
Double backticks around 5 literals in the above

anthonyeleven commented on src/mds/MDBalancer.cc:
Should the error string here and below include more detail, like perhaps an encoded representation of the string?

yongseokoh replied:
@anthonyeleven Could you please point me to the specific line of code you’re referring to?

Could you please confirm if this is the line you were referring to:
https://github.com/ceph/ceph/pull/52373/files#diff-729d5135082091929c032d8d6a0552bd2a5c658006aa7cb409764d85e64f9431R631

anthonyeleven replied:
I no longer see the line to which I was referring, so nevermind

anthonyeleven reviewed (commented):
Various nitpicky docs suggestions.

anthonyeleven commented on src/mds/MDBalancer.cc:
Odd that this doesn't appear to be anchored to the line in the file.

```
      dout(7) << "try to export nicely " << cd->get_path() << " auth " << cd->authority().first << " to " << target << " mask " << bitmask_to_str(rank_mask_bitset) << dendl;
      mds->mdcache->migrator->export_dir_nicely(cd, target);
```
What does it mean to export "nicely"?

yongseokoh replied:
@anthonyeleven 
“nicely” means performing a graceful export — the directory is transferred to the target MDS without forcing or interrupting ongoing operations.
I didn’t modify this function in this change.

yongseokoh commented on doc/cephfs/multimds.rst:
```diff
@@ -314,3 +313,151 @@ the ``bal_rank_mask`` should be set to ``0x0``. For example:
+can be overridden by configuring the ``ceph.dir.bal.mask`` vxattr.
+Notably, the ``ceph.dir.bal.mask`` offers the capability to isolate particular subdirectories
+into dedicated groups of MDS ranks, whereas bal_rank_mask only allows isolation
+at the file system level, such as the root directory ``/``. It serves as a valuable
```
Updated as suggested

yongseokoh commented on doc/cephfs/multimds.rst:
```diff
@@ -314,3 +313,151 @@ the ``bal_rank_mask`` should be set to ``0x0``. For example:
+        to various directories such as /usr/share and /mnt/cephfs.
+
+While the file system setting ``bal_rank_mask`` isolates the entire ``/`` directory to
+specific ranks, it can affect performance due to each other's migration overhead.
```
Updated as suggested

yongseokoh commented on doc/cephfs/multimds.rst:
```diff
@@ -314,3 +313,151 @@ the ``bal_rank_mask`` should be set to ``0x0``. For example:
+    setfattr -n ceph.dir.bal.mask -v 1,2 /usr/share/images
+    setfattr -n ceph.dir.bal.mask -v 3,4 /usr/share/backups
+
+.. note:: The balance_automate setting must be `true` for `ceph.dir.bal.mask` to have any effect.
```
Updated as suggested

yongseokoh commented on doc/cephfs/multimds.rst:
```diff
@@ -314,3 +313,151 @@ the ``bal_rank_mask`` should be set to ``0x0``. For example:
+
+.. note:: The balance_automate setting must be `true` for `ceph.dir.bal.mask` to have any effect.
+
+``/usr/share/images`` and ``/usr/share/backups`` are distributed within mds rank 1~2 and 3~4, respectively.
```
Updated as suggested

yongseokoh commented on doc/cephfs/multimds.rst:
```diff
@@ -314,3 +313,151 @@ the ``bal_rank_mask`` should be set to ``0x0``. For example:
+``/usr/share/images`` and ``/usr/share/backups`` are distributed within mds rank 1~2 and 3~4, respectively.
+
+Similar to ``ceph.dir.pin``, the ``ceph.dir.bal.mask`` is also inherited from its closest parent.
+This manner involves configuring the ``ceph.dir.bal.mask`` for a directory, which consequently
```
Updated as suggested

yongseokoh commented on doc/cephfs/multimds.rst:
```diff
@@ -314,3 +313,151 @@ the ``bal_rank_mask`` should be set to ``0x0``. For example:
+    # The subtree is now moved to mds rank 0.
+
+**Restrictions**: Since the inode of the root directory is always implicitly pinned to MDS rank 0,
+it's required to include MDS rank 0 when adjusting the ceph.dir.rank.mask for the ``/`` directory.
```
Updated as suggested

yongseokoh commented on qa/tasks/cephfs/test_exports.py:
```diff
@@ -632,185 +629,366 @@ def test_ephemeral_pin_shrink_mds(self):
 
-        self.mount_a.run_shell_payload('mkdir -p test/export')
+        if self.fs.get_var("max_mds") < 3:
+            self.skipTest("Require three mdss")
```
Updated as suggested

anthonyeleven reviewed (commented):
docs lgtm

yongseokoh commented on src/mds/MDBalancer.cc:
Updated as suggested

yongseokoh commented:
@anthonyeleven Changes applied as suggested. Let me know if I can rebase now.

yongseokoh committed:
mds: add ceph.dir.bal.mask vxattr for MDS Balancer

That introduces the ceph.dir.bal.mask vxattr, which
is an option to rebalance a subtree within
specific active MDSs. Similar to the CPU mask,
this feature enables load balancing of specific
directories across multiple MDS ranks. It is especially
useful for fine-tuning and improving performance
in various scenarios. Previously, the bal_rank_mask
in #43284 supports isolating unpinned subtrees under
the root directory ('/') to a specific MDS rank.
However, with this new option vxattr, it becomes
possible to isolate specific subdirectories to
designated MDS ranks. By introducing the
ceph.dir.bal.mask vxattr, this PR empowers
Ceph administrators with enhanced control and
flexibility for optimizing performance and
fine-tuning their deployments.

trakcer: https://tracker.ceph.com/issues/61777
Signed-off-by: Yongseok Oh <yongseok.oh@linecorp.com>

ceph-jenkins commented:
Can one of the admins verify this patch?

ceph-jenkins commented:
Thank you for your contribution.  Since you, the author, are not a member of the Ceph GitHub Org yet, our CI will not automatically run.  Any member of the Ceph Org may comment "ok  - to - test" (without the dashes) to allow the Jenkins jobs to run.

ceph-jenkins commented:
Thank you for your contribution.  Since you, the author, are not a member of the Ceph GitHub Org yet, our CI will not automatically run.  Any member of the Ceph Org may comment "ok  - to - test" (without the dashes) to allow the Jenkins jobs to run.

ceph-jenkins commented:
Thank you for your contribution.  Since you, the author, are not a member of the Ceph GitHub Org yet, our CI will not automatically run.  Any member of the Ceph Org may comment "ok  - to - test" (without the dashes) to allow the Jenkins jobs to run.

ceph-jenkins commented:
Thank you for your contribution.  Since you, the author, are not a member of the Ceph GitHub Org yet, our CI will not automatically run.  Any member of the Ceph Org may comment "ok  - to - test" (without the dashes) to allow the Jenkins jobs to run.

ceph-jenkins commented:
Thank you for your contribution.  Since you, the author, are not a member of the Ceph GitHub Org yet, our CI will not automatically run.  Any member of the Ceph Org may comment "ok  - to - test" (without the dashes) to allow the Jenkins jobs to run.

ceph-jenkins commented:
Thank you for your contribution.  Since you, the author, are not a member of the Ceph GitHub Org yet, our CI will not automatically run.  Any member of the Ceph Org may comment "ok  - to - test" (without the dashes) to allow the Jenkins jobs to run.

ceph-jenkins commented:
Thank you for your contribution.  Since you, the author, are not a member of the Ceph GitHub Org yet, our CI will not automatically run.  Any member of the Ceph Org may comment "ok  - to - test" (without the dashes) to allow the Jenkins jobs to run.

ceph-jenkins commented:
Thank you for your contribution.  Since you, the author, are not a member of the Ceph GitHub Org yet, our CI will not automatically run.  Any member of the Ceph Org may comment "ok  - to - test" (without the dashes) to allow the Jenkins jobs to run.

ceph-jenkins commented:
Thank you for your contribution.  Since you, the author, are not a member of the Ceph GitHub Org yet, our CI will not automatically run.  Any member of the Ceph Org may comment "ok  - to - test" (without the dashes) to allow the Jenkins jobs to run.

ceph-jenkins commented:
Thank you for your contribution.  Since you, the author, are not a member of the Ceph GitHub Org yet, our CI will not automatically run.  Any member of the Ceph Org may comment "ok  - to - test" (without the dashes) to allow the Jenkins jobs to run.

ceph-jenkins commented:
Thank you for your contribution.  Since you, the author, are not a member of the Ceph GitHub Org yet, our CI will not automatically run.  Any member of the Ceph Org may comment "ok  - to - test" (without the dashes) to allow the Jenkins jobs to run.

ceph-jenkins commented:
Thank you for your contribution.  Since you, the author, are not a member of the Ceph GitHub Org yet, our CI will not automatically run.  Any member of the Ceph Org may comment "ok  - to - test" (without the dashes) to allow the Jenkins jobs to run.

ceph-jenkins commented:
Thank you for your contribution.  Since you, the author, are not a member of the Ceph GitHub Org yet, our CI will not automatically run.  Any member of the Ceph Org may comment "ok  - to - test" (without the dashes) to allow the Jenkins jobs to run.

ceph-jenkins commented:
Thank you for your contribution.  Since you, the author, are not a member of the Ceph GitHub Org yet, our CI will not automatically run.  Any member of the Ceph Org may comment "ok  - to - test" (without the dashes) to allow the Jenkins jobs to run.

ceph-jenkins commented:
Thank you for your contribution.  Since you, the author, are not a member of the Ceph GitHub Org yet, our CI will not automatically run.  Any member of the Ceph Org may comment "ok  - to - test" (without the dashes) to allow the Jenkins jobs to run.

anthonyeleven commented:
This seems really useful, would love for someone to pick it back up

vshankar commented:
> This seems really useful, would love for someone to pick it back up

Yes, for Vampire 👍
